explorer/src/providers/cache.tsx

```diff
@@ -94,6 +95,15 @@ export function reducer<T, U>(
     case ActionType.Update: {
       const key = action.key;
       const entry = state.entries[key];
+
+      if (entry && entry.attempts === undefined) {
+        entry.attempts = 0;
```

Comment:
I don't think this code in particular would cause issues but typically data in reducers should always be immutable. So rather than directly mutating the entry, you could create a shallow copy like:
```js
const newEntry = {...entry};
```
and directly mutate that.

The pattern is important to avoid bugs in hook dependencies and component props. Imagine you use `entry.attempts` in a hook dependency array and the next render, React checks if `entry.attempts` changed but detects that it stayed the same because the original entry `attempts` field was updated to match the new one.

Reply:
Right right! Thanks for that tip! My previous attempt was immutable but I wasn't a fan of the ternaries.